curl/tests/mock_curl.cc

```diff
@@ -0,0 +1,361 @@
+            h->cacert = va_arg(vl, char*);
+            break;
+        case CURLOPT_SSLCERT:
+            // Set the mock curl SSL client cert name to that which was passed in the request.
+            if (h->test_failure_mode == curl_impl::error_mode::ssl_cert_error) {
+                va_end(vl);
+                return CURLE_COULDNT_CONNECT;
+            }
+            h->client_cert = va_arg(vl, char*);
+            break;
+        case CURLOPT_SSLKEY:
+            // Set the mock curl private keyfile name to that which was passed in the request.
+            if (h->test_failure_mode == curl_impl::error_mode::ssl_key_error) {
+                va_end(vl);
+                return CURLE_COULDNT_CONNECT;
+            }
+            h->client_key = va_arg(vl, char*);
+            break;
+        case CURLOPT_CONNECTTIMEOUT_MS:
+            if (h->test_failure_mode == curl_impl::error_mode::connect_timeout_error) {
+                va_end(vl);
+                return CURLE_COULDNT_CONNECT;
+            }
+            break;
+        case CURLOPT_TIMEOUT_MS:
+            if (h->test_failure_mode == curl_impl::error_mode::request_timeout_error) {
+                va_end(vl);
+                return CURLE_COULDNT_CONNECT;
+            }
+            break;
+        default:
+            break;
+    }
+    va_end(vl);
+    return CURLE_OK;
+}
+
+/*
+ * Perform a cURL transfer. the mock implementation uses this method to
+ * write the response header and body.
+ */
+CURLcode curl_easy_perform(CURL * easy_handle)
+{
+    auto h = static_cast<curl_impl*>(easy_handle);
+    if (h->test_failure_mode == curl_impl::error_mode::easy_perform_error) {
+        return CURLE_COULDNT_CONNECT;
+    }
+
+    /*
+     * Fill the read buffer in multiple chunks to better simulate real libcurl.
+     */
+    if (h->read_function) {
+        size_t bytes_returned;
+        char buf[10] = {};
+
+        while ((bytes_returned = h->read_function(buf, 1, 10, h->read_data))) {
+            h->read_buffer.append(buf, bytes_returned);
+        }
+    }
+
+    /*
+     * If we pass 'valid.com' in the test, return HTTP status 200. Otherwise, return status 404.
+     */
+    if (h->write_header) {
+        if (h->request_url == "http://valid.com/") {
+            string header_content = "HTTP/1.1 200 OK\n"
+                                    "Connection: keep-alive\n"
+                                    "Date: Thu, 16 Jul 2015 18:41:08 GMT\n"
+                                    "Content-Type: text/html;charset=UTF-8\n"
+                                    "Server: Jetty(7.x.y-SNAPSHOT)\n"
+                                    "Via: 1.1 vegur";
+
+            h->write_header(&header_content[0], 1, header_content.size(), h->header_context);
+        } else if (h->request_url == "http://nonstd-header.com/") {
+            string header_content = "nonstd_header_name:nonstd_header_value";
+            h->write_header(&header_content[0], 1, header_content.size(), h->header_context);
```

Comment:
I wonder if curl would really return CURLE_OK if it got these headers back. Maybe it doesn't care and just acts as a transport.